File: examples/rust/src/bin/masked_aggregation_operator.rs
use tsl_generated::tsl_core::StaticSimdVector;
use tsl_generated::tsl_scalar as tsl;

struct LessThan;

impl<V> tsl::algo::BinaryPredicateKernel<V> for LessThan
where
    V: StaticSimdVector<BaseType = i32> + tsl::detail::primitives::Less_thanImpl,
{
    fn test(&mut self, left: V::RegisterType, right: V::RegisterType) -> V::MaskType {
        tsl::less_than::<V>(left, right)
    }
}

struct MaskedSumOp {
    total: i64,
}

impl<V> tsl::algo::MaskedUnaryAggregateKernel<V> for MaskedSumOp
where
    V: StaticSimdVector<BaseType = i32>
        + tsl::detail::primitives::BlendImpl
        + tsl::detail::primitives::HaddImpl
        + tsl::detail::primitives::Set1Impl,
{
    type Output = i64;

    fn accumulate(&mut self, active: V::MaskType, value: V::RegisterType) {
        let zero = tsl::set1::<V>(0);
        let selected = tsl::blend::<V>(active, zero, value);
        self.total += i64::from(tsl::hadd::<V>(selected));
    }

    fn finalize(&self) -> Self::Output {
        self.total
    }
}

struct MaskedPairSumOp {
    total: i64,
}

impl<V> tsl::algo::MaskedBinaryAggregateKernel<V> for MaskedPairSumOp
where
    V: StaticSimdVector<BaseType = i32>
        + tsl::detail::primitives::AddImpl
        + tsl::detail::primitives::BlendImpl
        + tsl::detail::primitives::HaddImpl
        + tsl::detail::primitives::Set1Impl,
{
    type Output = i64;

    fn accumulate(&mut self, active: V::MaskType, left: V::RegisterType, right: V::RegisterType) {
        let zero = tsl::set1::<V>(0);
        let sum = tsl::add::<V>(left, right);
        let selected = tsl::blend::<V>(active, zero, sum);
        self.total += i64::from(tsl::hadd::<V>(selected));
    }

    fn finalize(&self) -> Self::Output {
        self.total
    }
}

fn fill_inputs(left: &mut [i32], right: &mut [i32]) {
    for (i, (left_value, right_value)) in left.iter_mut().zip(right.iter_mut()).enumerate() {
        *left_value = ((i * 7) % 41) as i32 - 20;
        *right_value = ((i * 13) % 59) as i32 - 29;
    }
}

fn expected_masked_sum(left: &[i32], right: &[i32]) -> i64 {
    left.iter()
        .zip(right)
        .filter_map(|(&left_value, &right_value)| {
            (left_value < right_value).then_some(i64::from(left_value))
        })
        .sum()
}

fn expected_masked_pair_sum(left: &[i32], right: &[i32]) -> i64 {
    left.iter()
        .zip(right)
        .filter_map(|(&left_value, &right_value)| {
            (left_value < right_value).then_some(i64::from(left_value) + i64::from(right_value))
        })
        .sum()
}

fn main() {
    let mut left = vec![0i32; 1003];
    let mut right = vec![0i32; 1003];
    fill_inputs(&mut left, &mut right);

    macro_rules! run_policy {
        ($policy:expr) => {{
            let policy = $policy;
            let mask_count = tsl::algo::integral_mask_chunk_count::<_, i32>(policy, left.len());
            let mut masks = vec![0u64; mask_count];
            let mut less_than = LessThan;
            let produced =
                tsl::algo::predicate_binary(policy, &mut less_than, &left, &right, &mut masks);
            assert_eq!(produced, masks.len());

            let mut unary = MaskedSumOp { total: 0 };
            let unary_result = tsl::algo::aggregate_masked_unary(policy, &mut unary, &left, &masks);
            assert_eq!(unary_result, expected_masked_sum(&left, &right));

            let mut binary = MaskedPairSumOp { total: 0 };
            let binary_result =
                tsl::algo::aggregate_masked_binary(policy, &mut binary, &left, &right, &masks);
            assert_eq!(binary_result, expected_masked_pair_sum(&left, &right));
        }};
    }

    run_policy!(tsl::algo::parallelism::native());
    run_policy!(tsl::algo::parallelism::fixed::<1>());
    run_policy!(tsl::algo::parallelism::generic::<4>());
    run_policy!(tsl::algo::parallelism::generic::<16>());
}

File: examples/cpp/masked_aggregation_operator.cpp
#include <cstddef>
#include <cstdint>
#include <memory>
#include <vector>

#include <tsl.hpp>

struct less_than_op {
    template <class Vec>
    typename Vec::mask_type operator()(
        typename tsl::reg_param<Vec>::type left,
        typename tsl::reg_param<Vec>::type right) const {
        return tsl::less_than<Vec>(left, right);
    }
};

struct masked_sum_op {
    std::int64_t total = 0;

    template <class Vec>
    void operator()(
        typename Vec::mask_type active,
        typename tsl::reg_param<Vec>::type value) {
        const auto zero = tsl::set1<Vec>(static_cast<typename Vec::base_type>(0));
        const auto selected = tsl::select<Vec>(active, value, zero);
        total += static_cast<std::int64_t>(tsl::hadd<Vec>(selected));
    }

    std::int64_t finalize() const {
        return total;
    }
};

struct masked_pair_sum_op {
    std::int64_t total = 0;

    template <class Vec>
    void operator()(
        typename Vec::mask_type active,
        typename tsl::reg_param<Vec>::type left,
        typename tsl::reg_param<Vec>::type right) {
        const auto zero = tsl::set1<Vec>(static_cast<typename Vec::base_type>(0));
        const auto sum = tsl::add<Vec>(left, right);
        const auto selected = tsl::select<Vec>(active, sum, zero);
        total += static_cast<std::int64_t>(tsl::hadd<Vec>(selected));
    }

    std::int64_t finalize() const {
        return total;
    }
};

void fill_inputs(
    std::vector<std::int32_t>& left,
    std::vector<std::int32_t>& right) {
    for (std::size_t i = 0; i < left.size(); ++i) {
        left[i] = static_cast<std::int32_t>(static_cast<int>((i * 7) % 41) - 20);
        right[i] = static_cast<std::int32_t>(static_cast<int>((i * 13) % 59) - 29);
    }
}

std::int64_t expected_masked_sum(
    const std::vector<std::int32_t>& left,
    const std::vector<std::int32_t>& right) {
    std::int64_t total = 0;
    for (std::size_t i = 0; i < left.size(); ++i) {
        if (left[i] < right[i]) {
            total += left[i];
        }
    }
    return total;
}

std::int64_t expected_masked_pair_sum(
    const std::vector<std::int32_t>& left,
    const std::vector<std::int32_t>& right) {
    std::int64_t total = 0;
    for (std::size_t i = 0; i < left.size(); ++i) {
        if (left[i] < right[i]) {
            total += static_cast<std::int64_t>(left[i]) + right[i];
        }
    }
    return total;
}

template <class MaskLayout, class Parallelism>
bool run_masked_aggregation_cases() {
    constexpr std::size_t count = 1003;
    std::vector<std::int32_t> left(count);
    std::vector<std::int32_t> right(count);
    fill_inputs(left, right);

    using mask_type =
        tsl::algo::mask_storage_type<MaskLayout, Parallelism, std::int32_t>;
    const auto mask_count =
        tsl::algo::mask_chunk_count<MaskLayout, Parallelism, std::int32_t>(count);
    std::unique_ptr<mask_type[]> masks(new mask_type[mask_count]);

    const auto produced = tsl::algo::predicate_binary<
        Parallelism,
        tsl::algo::alignment::unaligned,
        MaskLayout>(
        less_than_op{},
        left.data(),
        right.data(),
        masks.get(),
        count);
    if (produced != mask_count) {
        return false;
    }

    const auto unary_result = tsl::algo::aggregate_masked_unary<
        Parallelism,
        tsl::algo::alignment::unaligned,
        MaskLayout>(
        masked_sum_op{},
        left.data(),
        masks.get(),
        count);
    if (unary_result != expected_masked_sum(left, right)) {
        return false;
    }

    const auto binary_result = tsl::algo::aggregate_masked_binary<
        Parallelism,
        tsl::algo::alignment::unaligned,
        MaskLayout>(
        masked_pair_sum_op{},
        left.data(),
        right.data(),
        masks.get(),
        count);
    return binary_result == expected_masked_pair_sum(left, right);
}

template <class Parallelism>
bool run_layout_cases() {
    return run_masked_aggregation_cases<
               tsl::algo::mask_layout::integral,
               Parallelism>() &&
           run_masked_aggregation_cases<
               tsl::algo::mask_layout::native,
               Parallelism>() &&
           run_masked_aggregation_cases<
               tsl::algo::mask_layout::bytes,
               Parallelism>() &&
           run_masked_aggregation_cases<
               tsl::algo::mask_layout::bits,
               Parallelism>();
}

int main() {
    if (!run_layout_cases<tsl::dataparallel::fixed<1>>()) {
        return 1;
    }
    if (!run_layout_cases<tsl::dataparallel::generic<4>>()) {
        return 2;
    }
    if (!run_layout_cases<tsl::dataparallel::generic<16>>()) {
        return 3;
    }
    return 0;
}
